File: composite/typescript/product.ts
import {ICatalogComponent} from "./catalog-component.interface";

// Somut yapımızdır.
// UML diyagramındaki Leaf kısmına denk gelmektedir.
// Hiyerarşinin en alt tabakasını temsil etmektedir.
export class Product implements ICatalogComponent {
  private readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  public drawHierarchy(): void {
    console.log(this.name);
  }
}

File: composite/go/product.go
package main

import "fmt"

// Somut yapımızdır.
// UML diyagramındaki Leaf kısmına denk gelmektedir.
// Hiyerarşinin en alt tabakasını temsil etmektedir.
type product struct {
	name string
}

func NewProduct(name string) *product {
	return &product{name: name}
}

func (p *product) DrawHierarchy() {
	fmt.Println(p.name)
}
